### PR TITLE
Move userId to controller to fix tagging despite option disabled

### LINKED ISSUE #5290: Mails are getting automatically classified even if classification is disabled
### Expected behavior
After #5270 was merged, I was expecting that deselecting that new setting would stop Nextcloud from making virtually all new Mails show up as Important in Thunderbird.

### Actual behavior
Most new mails are still getting Tagged as important, despite the setting being disabled.

### Mail app

**Mail app version:** 1.10.0

**Mailserver or service:** Dovecot

### Server configuration

**Operating system**: Gentoo

**Web server:** Apache

**Database:** PostgreSQL

**PHP version:** 7.4

**Nextcloud Version:** 22.0.0

#### Client configuration
**Browser:** Edge 91

**Operating system:** Windows

### PR DESCRIPTION
**Pull request by miaulalala**
Cron runs did still tag messages because the user ID injection does not work for cron.

Fixes #5290

### CODE REVIEW

File: lib/Contracts/IUserPreferences.php [hunk 1]
```diff
@@ -29,15 +29,18 @@
...
 
 	/**
+	 * @param string $userId
 	 * @param string $key
 	 * @param mixed $value
 	 * @return mixed new value
+	 * @throws \OCP\PreConditionNotMetException
 	 */
-	public function setPreference($key, $value);
+	public function setPreference($userId, $key, $value);
```
Comment (ChristophWurst) [inline lib/Contracts/IUserPreferences.php:38]:
```suggestion
	public function setPreference(string $userId, $key, $value);
```

File: lib/Contracts/IUserPreferences.php [hunk 2]
```diff
@@ -29,15 +29,18 @@
...
 
 	/**
+	 * @param string $userId
 	 * @param string $key
 	 * @param mixed|null $default
 	 */
-	public function getPreference($key, $default = null);
+	public function getPreference($userId, $key, $default = null);
```
Comment (ChristophWurst) [inline lib/Contracts/IUserPreferences.php:45]:
```suggestion
	public function getPreference(string $userId, $key, $default = null);
```

File: lib/Controller/PreferencesController.php [hunk 3]
```diff
@@ -31,20 +31,26 @@
...
 use OCP\IRequest;
+use OCP\PreConditionNotMetException;
 
 class PreferencesController extends Controller {
 
 	/** @var IUserPreferences */
 	private $userPreference;
 
+	/** @var string */
+	private $UserId;
```
Comment (ChristophWurst) [inline lib/Controller/PreferencesController.php:42]:
```suggestion
	private $userId;
```

capitalization on the first letter is only required for the argument on the constructor. The internal member we can case correctly :)

**Review by ChristophWurst — APPROVED** [review]

File: lib/Controller/PreferencesController.php (hunk 3 above)
Comment (miaulalala) [reply lib/Controller/PreferencesController.php:42]: Ah good to know! Will change :)

File: lib/Contracts/IUserPreferences.php [hunk 4]
```diff
@@ -29,15 +29,18 @@
 interface IUserPreferences {
 
 	/**
+	 * @param string $userId
 	 * @param string $key
 	 * @param mixed $value
 	 * @return mixed new value
+	 * @throws \OCP\PreConditionNotMetException
```
Comment (kesselb) [inline lib/Contracts/IUserPreferences.php:36]: We don't pass $preCondition to setUserValue so the method will not throw a PreConditionNotMetException. I would prefer to not list in phpdocs.

**Comment by kesselb** [comment]
Needs an update:

- https://github.com/nextcloud/mail/blob/267f964aa37c292dc1551d6e7687942e5eec2749/lib/Listener/AddressCollectionListener.php#L59
- https://github.com/nextcloud/mail/blob/0a1e1ad8bc46bb4d4b0846b840cd6b8a4ff3c408/lib/Service/AvatarService.php#L81

File: lib/Controller/PreferencesController.php [hunk 5]
```diff
@@ -74,7 +80,11 @@ public function update($key, $value): JSONResponse {
 			throw new ClientException('key or value missing');
 		}
 
-		$newValue = $this->userPreference->setPreference($key, $value);
+		try {
+			$newValue = $this->userPreference->setPreference($this->userId, $key, $value);
+		} catch (PreConditionNotMetException $e) {
+			throw new ClientException('key or value invalid');
```
Comment (kesselb) [inline lib/Controller/PreferencesController.php:86]: Please remove try-catch here. The method will not throw a PreConditionNotMetException for us.

File: lib/Controller/PreferencesController.php [hunk 6]
```diff
@@ -31,20 +31,26 @@
 use OCP\AppFramework\Controller;
 use OCP\AppFramework\Http\JSONResponse;
 use OCP\IRequest;
+use OCP\PreConditionNotMetException;
```
Comment (kesselb) [inline lib/Controller/PreferencesController.php:34]:
```suggestion
```

**Comment by ChristophWurst** [comment]
please rebase and squash

**Review by ChristophWurst — APPROVED** [review]

**Pull request merged**